### PR TITLE
Add label to grades which have changed since last sync

### PR DESCRIPTION
**Pull request by acelaya**
> Depends on https://github.com/hypothesis/lms/pull/6710

Indicate which students have a new auto-calculated grade since the last time grades were synced.

Additionally, when syncing grades, exclude those which didn't change since last sync.

image

image

### Considerations

1. For now, I decided to remove the "New" label from students once the sync has been scheduled, without waiting for the actual sync to finish.
    I did this because it was simpler than waiting for the actual sync to finish, but once we have sync statuses per student, I will refactor the logic to that it is more accurate and is not "broken" if the pages is reloaded or if the sync is already in progress when the page is loaded for the first time.

### TODO

- [x] After syncing students, remove the label from those which have been successfully synced.
- [x] Add tests.

### Testing steps

1. Checkout this branch
2. Delete all synced grades in your local DB: `truncate grading_sync cascade;`
3. Go to https://hypothesis.instructure.com/courses/319/assignments/7296, and edit the assignment so that it has auto-grading enabled.
4. In the sidebar, click on the user menu and then on "Open dashboard". You should see pre-calculated grades for all students, and all should have the "New" label next to them.
5. The sync button should be including all existing students. Click on it, and soon after clicking, all the "new" labels should be removed from students.
6. Wait for sync to finish, then reload the page. The students should still not show "New" labels, and the sync button should be disabled with the text "Grades synced".
7. Log in Canvas with a student account, and add some new annotation. If you come back to the dashboard, that student should have the "New" label next to its grade, and the grade popover should show their previously synced grade.
8. The sync button should be including only that student now.

### CODE REVIEW

File: lms/static/scripts/frontend_apps/components/dashboard/AssignmentActivity.tsx [hunk 1]
```diff
@@ -31,6 +31,7 @@ type StudentsTableRow = {
   annotations: number;
   replies: number;
   current_grade?: number;
+  last_grade?: number | null;
```
Comment (robertknight) [inline lms/static/scripts/frontend_apps/components/dashboard/AssignmentActivity.tsx:34]: Perhaps rename to `last_submitted_grade` to make it clear what point in time "last" refers to, or add JSDoc comments which explain this.

File: lms/static/scripts/frontend_apps/components/dashboard/GradeIndicator.tsx [hunk 2]
```diff
@@ -77,23 +85,40 @@ export default function GradeIndicator({
   const requiredCombined = config
     ? config.required_annotations + (config.required_replies ?? 0)
     : 0;
+  // Checking typeof lastGrade to avoid number zero to be treated as false
+  const hasLastGrade = typeof lastGrade === 'number';
+  const gradeHasChanged = lastGrade !== grade;
+  const showNewIndicator = !hasLastGrade || gradeHasChanged;
```
Comment (robertknight) [inline lms/static/scripts/frontend_apps/components/dashboard/GradeIndicator.tsx:91]: In the case where there is no last grade, I expect that `lastGrade` is null/undefined and `grade` is a number, so `gradeHasChanged` will be true, meaning `!hasLastGrade` is not strictly required?

File: lms/static/scripts/frontend_apps/components/dashboard/SyncGradesButton.tsx [hunk 3]
```diff
@@ -20,10 +20,18 @@ export type SyncGradesButtonProps = {
    * Passing `undefined` means students are not yet known and/or being loaded.
    */
   studentsToSync?: Array<{ h_userid: string; grade: number }>;
+
+  /**
+   * Invoked after a sync was scheduled successfully.
```
Comment (robertknight) [inline lms/static/scripts/frontend_apps/components/dashboard/SyncGradesButton.tsx:25]: What happens if the sync is not scheduled successfully? Is that scenario handled internally within the component?

**Review by robertknight — APPROVED** [review]
LGTM. I did encounter some issues in testing, discussed in [this thread](https://hypothes-is.slack.com/archives/C4K6M7P5E/p1727778716442739), but they are unrelated to the changes here.

File: lms/static/scripts/frontend_apps/components/dashboard/SyncGradesButton.tsx (hunk 3 above)
Comment (acelaya) [reply lms/static/scripts/frontend_apps/components/dashboard/SyncGradesButton.tsx:25]: > Is that scenario handled internally within the component?

Yes. In that case the last sync is marked as error, which will put the button in a "Sync failed. Click to retry" state.

https://github.com/hypothesis/lms/pull/6712/files#diff-581f6b4dde8f047b2629b55a1a0730a8509f31990880fb065ddfc193a283637eL117

File: lms/static/scripts/frontend_apps/components/dashboard/GradeIndicator.tsx (hunk 2 above)
Comment (acelaya) [reply lms/static/scripts/frontend_apps/components/dashboard/GradeIndicator.tsx:91]: Yep! You are right.

**Pull request merged**